File: snippets/dart_snippets/lib/pay_onchain.dart
import 'package:dart_snippets/sdk_instance.dart';
import 'package:flutter_breez_liquid/flutter_breez_liquid.dart';

Future<OnchainPaymentLimitsResponse> getCurrentLimits() async {
  // ANCHOR: get-current-pay-onchain-limits
  OnchainPaymentLimitsResponse currentLimits = await breezSDKLiquid.instance!.fetchOnchainLimits();
  print("Minimum amount: ${currentLimits.send.minSat} sats");
  print("Maximum amount: ${currentLimits.send.maxSat} sats");
  // ANCHOR_END: get-current-pay-onchain-limits
  return currentLimits;
}

Future<PreparePayOnchainResponse> preparePayOnchain() async {
  // ANCHOR: prepare-pay-onchain
  PreparePayOnchainRequest preparePayOnchainRequest = PreparePayOnchainRequest(
    amount: PayAmount_Bitcoin(receiverAmountSat: 5000 as BigInt),
  );
  PreparePayOnchainResponse prepareRes = await breezSDKLiquid.instance!.preparePayOnchain(
    req: preparePayOnchainRequest,
  );

  // Check if the fees are acceptable before proceeding
  BigInt totalFeesSat = prepareRes.totalFeesSat;
  // ANCHOR_END: prepare-pay-onchain
  print(totalFeesSat);
  return prepareRes;
}

Future<PreparePayOnchainResponse> preparePayOnchainDrain() async {
  // ANCHOR: prepare-pay-onchain-drain
  PreparePayOnchainRequest preparePayOnchainRequest = PreparePayOnchainRequest(
    amount: PayAmount_Drain(),
  );
  PreparePayOnchainResponse prepareRes = await breezSDKLiquid.instance!.preparePayOnchain(
    req: preparePayOnchainRequest,
  );

  // Check if the fees are acceptable before proceeding
  BigInt totalFeesSat = prepareRes.totalFeesSat;
  // ANCHOR_END: prepare-pay-onchain-drain
  print(totalFeesSat);
  return prepareRes;
}

Future<PreparePayOnchainResponse> preparePayOnchainFeeRate() async {
  // ANCHOR: prepare-pay-onchain-fee-rate
  int optionalSatPerVbyte = 21;

  PreparePayOnchainRequest preparePayOnchainRequest = PreparePayOnchainRequest(
    amount: PayAmount_Bitcoin(receiverAmountSat: 5000 as BigInt),
    feeRateSatPerVbyte: optionalSatPerVbyte,
  );
  PreparePayOnchainResponse prepareRes = await breezSDKLiquid.instance!.preparePayOnchain(
    req: preparePayOnchainRequest,
  );

  // Check if the fees are acceptable before proceeding
  BigInt claimFeesSat = prepareRes.claimFeesSat;
  BigInt totalFeesSat = prepareRes.totalFeesSat;
  // ANCHOR_END: prepare-pay-onchain-fee-rate
  print(claimFeesSat);
  print(totalFeesSat);
  return prepareRes;
}

Future<SendPaymentResponse> startReverseSwap({
  required PreparePayOnchainResponse prepareRes,
}) async {
  // ANCHOR: start-reverse-swap
  String destinationAddress = "bc1..";

  PayOnchainRequest req = PayOnchainRequest(
    address: destinationAddress,
    prepareResponse: prepareRes,
  );
  SendPaymentResponse res = await breezSDKLiquid.instance!.payOnchain(req: req);
  // ANCHOR_END: start-reverse-swap
  return res;
}

File: snippets/wasm/pay_onchain.ts
import { type BindingLiquidSdk, type PreparePayOnchainResponse } from '@breeztech/breez-sdk-liquid'

const exampleGetCurrentLimits = async (sdk: BindingLiquidSdk) => {
  // ANCHOR: get-current-pay-onchain-limits
  try {
    const currentLimits = await sdk.fetchOnchainLimits()

    console.log(`Minimum amount, in sats: ${currentLimits.send.minSat}`)
    console.log(`Maximum amount, in sats: ${currentLimits.send.maxSat}`)
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: get-current-pay-onchain-limits
}

const examplePreparePayOnchain = async (sdk: BindingLiquidSdk) => {
  // ANCHOR: prepare-pay-onchain
  try {
    const prepareResponse = await sdk.preparePayOnchain({
      amount: {
        type: 'bitcoin',
        receiverAmountSat: 5_000
      }
    })

    // Check if the fees are acceptable before proceeding
    const totalFeesSat = prepareResponse.totalFeesSat
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: prepare-pay-onchain
}

const examplePreparePayOnchainDrain = async (sdk: BindingLiquidSdk) => {
  // ANCHOR: prepare-pay-onchain-drain
  try {
    const prepareResponse = await sdk.preparePayOnchain({
      amount: {
        type: 'drain'
      }
    })

    // Check if the fees are acceptable before proceeding
    const totalFeesSat = prepareResponse.totalFeesSat
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: prepare-pay-onchain-drain
}

const examplePreparePayOnchainFeeRate = async (sdk: BindingLiquidSdk) => {
  // ANCHOR: prepare-pay-onchain-fee-rate
  try {
    const optionalSatPerVbyte = 21

    const prepareResponse = await sdk.preparePayOnchain({
      amount: {
        type: 'bitcoin',
        receiverAmountSat: 5_000
      },
      feeRateSatPerVbyte: optionalSatPerVbyte
    })

    // Check if the fees are acceptable before proceeding
    const claimFeesSat = prepareResponse.claimFeesSat
    const totalFeesSat = prepareResponse.totalFeesSat
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: prepare-pay-onchain-fee-rate
}

const examplePayOnchain = async (
  sdk: BindingLiquidSdk,
  prepareResponse: PreparePayOnchainResponse
) => {
  // ANCHOR: start-reverse-swap
  try {
    const destinationAddress = 'bc1..'

    const payOnchainRes = await sdk.payOnchain({
      address: destinationAddress,
      prepareResponse
    })
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: start-reverse-swap
}
